[Frontend] Add OpenAI Vision API Support

ywang96 opened the pull request:
This PR is the final step of the release milestone of [adding support for GPT4V compatible inference](https://github.com/vllm-project/vllm/issues/4194#issuecomment-2126436729).

This PR fixes 
- #2058
- #3873
- #4205
- #4826

and closes #4200

ywang96 committed:
initial

ywang96 committed:
iterate

ywang96 committed:
Merge branch 'main' into gpt4v-fe

ywang96 committed:
iterate

ywang96 committed:
iterate

ywang96 committed:
iterate

ywang96 committed:
iterate

ywang96 committed:
adding test
Co-Authored-By: Cyrus Leung <tlleungac@connect.ust.hk>

ywang96 committed:
iterate

ywang96 committed:
docstring

ywang96 committed:
remove unused lib

ywang96 commented:
@DarkLight1337 This PR is probably good for first pass - I will mark it as "Ready for review" after I address your first round of review, along with adding more test and documentation.

ywang96 committed:
revert hardcoded chat template

DarkLight1337 commented on tests/entrypoints/test_openai_vision.py:
```diff
@@ -0,0 +1,167 @@
+    "https://upload.wikimedia.org/wikipedia/commons/0/0b/RGBA_comp.png",
+]
+
+pytestmark = pytest.mark.asyncio
```
`pytestmark` is now used for indicating the group to test (refer to `test_openai_server.py`). Please make changes accordingly, otherwise the test won't be included in CI.

DarkLight1337 replied:
Hmm seems that the tests passed regardless.

DarkLight1337 commented on vllm/entrypoints/openai/serving_chat.py:
```diff
@@ -93,19 +98,75 @@ def _parse_chat_message_content_parts(
+
+                    image_future = async_get_and_parse_image(image_url["url"])
+                    image_futures.append(image_future)
+                    image_count += 1
```
Imo my original method of using `len(image_futures)` as the image count is less prone to error compared to having to increment the image count manually.

ywang96 replied:
I thought about keeping a counter in case we need it in the future when can support multiple images per request, but let's do it the simple way now.

DarkLight1337 commented on vllm/entrypoints/openai/serving_chat.py:
```diff
@@ -93,19 +98,75 @@ def _parse_chat_message_content_parts(
+        model_config = getattr(self.engine.engine, "model_config", None)
+
+        image_count = 0
         for _, part in enumerate(parts):
```
Can iterate over `parts` directly if we aren't using the index anymore.

DarkLight1337 commented on vllm/entrypoints/openai/serving_chat.py:
```diff
@@ -180,15 +254,27 @@ async def create_chat_completion(
-            request_id,
-            lora_request,
-        )
+        if self.engine.engine.vision_language_config is None:
```
There is quite a bit of repetition here. I have updated #4200 with a simpler version of this.

ywang96 replied:
Hmm...I guess `multi_modal_data` will always be `None` if `vision_language_config` is None, so yes we can simplify this.

DarkLight1337 commented on vllm/envs.py:
```diff
@@ -213,6 +214,11 @@
+    # Timeout for fetching images when serving multimodal models
+    # Default is 2 seconds
+    "VLLM_IMAGE_FETCH_TIMEOUT":
+    lambda: int(os.getenv("VLLM_IMAGE_FETCH_TIMEOUT", "2")),
```
A quick look at [the requests documentation](https://requests.readthedocs.io/en/latest/user/advanced/#timeouts) suggests that we should set the timeout to at least 3 seconds.

DarkLight1337 commented on vllm/multimodal/utils.py:
```diff
@@ -0,0 +1,61 @@
+    return img
+
+
+async_fetch_image = make_async(fetch_image)  # type: ignore
```
Why do we have to `type: ignore` this?

ywang96 replied:
was a typo

DarkLight1337 commented on vllm/multimodal/utils.py:
```diff
@@ -0,0 +1,61 @@
+        response.raise_for_status()
+
+        # Open the image using PIL
+        img = Image.open(BytesIO(response.content))
```
Is this more efficient than the streaming approach used in #4200?

ywang96 replied:
I believe this is the proper way to read an image data (albeit I haven't done any testing). See https://stackoverflow.com/questions/7391945/how-do-i-read-image-data-from-a-url-in-python

Since we're going to `await` this anyways, would streaming vs non-streaming matter here?

DarkLight1337 replied:
Based on [this requests documentation](https://requests.readthedocs.io/en/latest/user/quickstart/#binary-response-content), it seems you're right. Let's keep this code then.

DarkLight1337 commented on vllm/entrypoints/openai/serving_chat.py:
```diff
@@ -93,19 +98,75 @@ def _parse_chat_message_content_parts(
+            (image_token_prompt,
+             image_token_str) = vlm_config.get_image_token_text(self.tokenizer)
+
+            # NOTE: If image token string (e.g, <image>) is already present
```
IMO we should not allow this as it could open up the server to token injection (maliciously passing in incorrect number of image tokens, which is not validated before passing to the model) which may crash the async engine.

ywang96 replied:
I will put a warning here

DarkLight1337 reviewed (commented):
Thanks for the hard work! I have left a few comments.

ywang96 committed:
address feedback

ywang96 committed:
update pytestmark

DarkLight1337 commented on tests/entrypoints/test_openai_vision.py:
```diff
@@ -0,0 +1,167 @@
+                                                           max_tokens=10,
+                                                           logprobs=True,
+                                                           top_logprobs=5)
+    assert chat_completion.id is not None
```
This test is outdated, please see [57d65eb](https://github.com/vllm-project/vllm/pull/4200/commits/57d65eb00d7cf37fb07064a3832ce0da2439cbdc).

DarkLight1337 replied:
I found that the number of tokens should be as shown in [ddf3f06](https://github.com/vllm-project/vllm/pull/4200/commits/ddf3f06a76be6dcb90641f68419f0b7c7a98565f).

ywang96 committed:
apply asyncio mark

ywang96 committed:
update doc

ywang96 committed:
update test

ywang96 commented:
With docs and tests modified - going to make this PR ready for review cc @DarkLight1337.

DarkLight1337 commented:
Assuming that the tests pass, LGTM.

ywang96 committed:
minor doc update

DarkLight1337 commented:
Seems that the docs are formatted improperly.

ywang96 committed:
minor doc update

ywang96 committed:
Clarify experiment support

ywang96 committed:
note regarding prompt format when using API server

ywang96 commented:
> Seems that the docs are formatted improperly.

I just made a few changes to add some final edits - see https://vllm--5237.org.readthedocs.build/en/5237/

ywang96 committed:
Merge branch 'main' into gpt4v-fe

DarkLight1337 commented:
The docs look good now. Shall we merge this?

ywang96 commented:
> The docs look good now. Shall we merge this?

Thanks for the approval! Let's not rush this merge and I still want to add a few tests as well as running a few integration tests myself!

ywang96 commented:
Looks like entrypoint tests are actually failing - fixing them now..

ywang96 committed:
fix typo

ywang96 committed:
update template

ywang96 committed:
revert and update token count

ywang96 committed:
update template

ywang96 committed:
update

ywang96 committed:
update

ywang96 committed:
template format

ywang96 committed:
correct and add test for multi image

ywang96 committed:
fix test

ywang96 commented:
@DarkLight1337 I've fixed the chat template per our offline discussion, and added tests for base64 encoded images and  multi-image inputs. All the entrypoint tests have passed now.

Since this is a new user facing feature, can I ask one of you @simon-mo @zhuohan123 @WoosukKwon to also take a quick look and see if we're missing anything? Otherwise I'd appreciate a sign-off so we can merge this PR. Thanks!

DarkLight1337 committed:
Add unit test for `fetch_image`

DarkLight1337 commented:
As discussed offline, I have adapted the unit test for fetching images from #4200.

DarkLight1337 committed:
Apply formatter

simon-mo commented on vllm/config.py:
```diff
@@ -1114,6 +1114,14 @@ def get_image_input_enum_type(cls, value: str) -> ImageInputType:
                              f"Expecting to choose from "
                              f"{[x.name for x in cls.ImageInputType]}.") from e
 
+    def get_image_token_text(
```
the computed value can be cached, but it is also super simple to recompute.

ywang96 replied:
The way `VisionLanguageConfig` was designed in #3042 made this call have to depend on the tokenizer, which's less clean. I left a TODO on adding `@cached_property` on this call once we refactor `VisionLanguageConfig`

simon-mo commented on vllm/entrypoints/openai/serving_chat.py:
```diff
@@ -40,6 +45,7 @@ class ConversationMessage(TypedDict):
 @dataclass(frozen=True)
 class ChatMessageParseResult:
     messages: List[ConversationMessage]
+    image_futures: List[Awaitable[ImagePixelData]]
```
default factory list

DarkLight1337 replied:
I think Simon meant to use `dataclass.field` here, not inside the function.

simon-mo commented on vllm/envs.py:
```diff
@@ -213,6 +214,11 @@
+    # Timeout for fetching images when serving multimodal models
+    # Default is 3 seconds
+    "VLLM_IMAGE_FETCH_TIMEOUT":
+    lambda: int(os.getenv("VLLM_IMAGE_FETCH_TIMEOUT", "3")),
```
this is a bit too short for large image? maybe 5/10s?

ywang96 replied:
I will bring this to 5s for now, and add documentation that this can be configured by setting this env.

simon-mo commented on vllm/multimodal/utils.py:
```diff
@@ -0,0 +1,71 @@
+from typing import Union
+
+import requests
+from PIL import Image
```
can you double check this is in requirements txt

ywang96 replied:
This was added in the last image processor PR. https://github.com/vllm-project/vllm/blob/7b0a0dfb22907505441f8a4a5eb882cbca4d2acf/requirements-common.txt#L15

simon-mo commented on vllm/multimodal/utils.py:
```diff
@@ -0,0 +1,71 @@
+    return img
+
+
+async_fetch_image = make_async(fetch_image)
```
FYI the perf will be really bad here because it will be spawning a thread per image download (won't block the event loop but will consume Python GIL). It might be easier to introduce dependency for aiohttp or httpx and make it natively async so it's faster to benchmark.

ywang96 replied:
Sounds good, I will take a look and modify this.

simon-mo commented on vllm/multimodal/utils.py:
```diff
@@ -0,0 +1,71 @@
+    return Image.open(BytesIO(base64.b64decode(image)))
+
+
+def encode_image_base64(image: Image.Image, format: str = 'JPEG') -> str:
```
where is this used?

ywang96 replied:
This is added only for testing now, but I figured this could be a helpful util for the developer/user to encode their image so I didn't put it in the test file.

ywang96 committed:
address feedback

ywang96 committed:
fix notes

ywang96 committed:
use aiohttp

ywang96 committed:
fix test

ywang96 committed:
test

DarkLight1337 commented on tests/entrypoints/test_openai_vision.py:
```diff
@@ -117,7 +117,8 @@ async def test_single_chat_session_image(server, client: openai.AsyncOpenAI,
 
-    image_encoded = encode_image_base64(fetch_image(image_url=image_url))
+    image_encoded = encode_image_base64(
+        ImageFetchAiohttp.fetch_image(image_url=image_url))
```
Forgot to await this

ywang96 committed:
fix test

ywang96 committed:
update test

ywang96 committed:
update fixture

ywang96 committed:
fix field

DarkLight1337 commented on vllm/entrypoints/openai/serving_chat.py:
```diff
@@ -94,19 +100,78 @@ def _parse_chat_message_content_parts(
     ) -> ChatMessageParseResult:
         texts: List[str] = []
+        image_futures: List[Awaitable[ImagePixelData]] = Field(
+            default_factory=list)
```
See above

ywang96 replied:
I didn't read too carefully and looked at the wrong place...

ywang96 committed:
fix field

ywang96 committed:
format

ywang96 committed:
fix image loading

ywang96 committed:
revert change that merges fetch and parse

ywang96 committed:
add encoded image fixture

ywang96 committed:
Merge branch 'main' into gpt4v-fe

ywang96 committed:
update fetch image and remove unused fixture

ywang96 committed:
cleanup

ywang96 committed:
fix fixture

ywang96 committed:
remove unused client close

ywang96 commented:
Thanks for the review @simon-mo! I have addressed the comments and all related tests have passed. Could you take a look at the `aiohttp` implementation and let me know if there's any issue? Thanks!

ywang96 committed:
add TODO and format

DarkLight1337 commented on vllm/multimodal/utils.py:
```diff
@@ -23,30 +24,26 @@ def get_aiohttp_client(cls) -> aiohttp.ClientSession:
-            image = load_image_from_base64(image_url.split(',')[1])
+            loop = asyncio.get_event_loop()
+            image = await loop.run_in_executor(None, load_image_from_base64,
+                                               image_url.split(',')[1])
```
- I don't see how this way is different from using `make_async`.
- We should set a limit of `1` to `image_url.split` so it won't keep iterating through the string.

ywang96 replied:
- Good catch - I was testing if this had any perf implication by making it async but forgot to change it back.
- That's a good point about split for sure!

ywang96 committed:
address comment

simon-mo commented:
* can you add the package to requirements 
* can you confirm the test is ran in ci (the entrypoints tests config in pipeline is complex)

DarkLight1337 commented:
> * can you add the package to requirements

`aiohttp` was already added to common requirements in #4794.

> * can you confirm the test is ran in ci (the entrypoints tests config in pipeline is complex)

I have downloaded the CI log and can confirm this (with the help of `Ctrl-F`).
